compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -4212,6 +4213,68 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
           typedExpr(cmp, defn.BooleanType)
       case _ =>
 
+  // if expression in statement position (of template or block)
+  // looks like a useful value that should not be ignored, warn and return true
+  // User specifies that an expression is boring by ascribing `e: Unit`.
+  // The subtree `e` will bear an attachment, but may be wrapped in adaptations.
+  private def checkInterestingResultInStatement(t: Tree)(using Context): Boolean = {
+    def isUninterestingSymbol(sym: Symbol): Boolean =
+      sym == NoSymbol ||
+      sym.isConstructor ||
+      sym.is(Package) ||
+      sym.isPackageObject ||
+      sym.isClass ||
+      sym == defn.BoxedUnitClass ||
+      sym == defn.AnyClass ||
+      sym == defn.AnyRefAlias ||
+      sym == defn.AnyValClass
+    def isUninterestingType(tpe: Type): Boolean =
+      tpe == NoType ||
+      tpe.typeSymbol == defn.UnitClass ||
+      defn.isBottomClass(tpe.typeSymbol) ||
+      tpe =:= defn.UnitType ||
+      tpe.typeSymbol == defn.BoxedUnitClass ||
+      tpe =:= defn.AnyValType ||
+      tpe =:= defn.AnyType ||
+      tpe =:= defn.AnyRefType
+    // java lacks this.type idiom to distinguish side-effecting method, so ignore result of invoking java method.
+    def isJavaApplication(t: Tree): Boolean = t match {
+      case Apply(f, _) => f.symbol.is(JavaDefined) && !defn.ObjectClass.isSubClass(f.symbol.owner)
+      case _ => false
+    }
+    // The quirk of typechecking if is that the LUB often results in boring types.
+    // Parser adds suppressing attachment on `if (b) expr` when user has `-Wnonunit-if:false`.
```

Comment:
We don't have `-Wnonunit-if` setting so this comment doesn't seem to be true

Reply:
Right, removed the irrelevant part of the comment. Thanks